### PR TITLE
Smorgrav/cost api take 2

### PR DESCRIPTION
**Pull request by smorgrav**
@ldalves FYI Early draft...

### CODE REVIEW

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/Cost.java [hunk 1]
```diff
@@ -2,52 +2,110 @@
...
+            });
+
+        });
+
+        return costApplications;
+    }
 
     /**
-     * Collect all information and format it as a Cvs blob for download.
+     * Get application costs for a specific application deployement
```
Comment (ldalves) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/Cost.java]: s/deployement/deployment/

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/CostApplication.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+package com.yahoo.vespa.hosted.controller.api.integration.cost;
+
+import com.yahoo.config.provision.ApplicationId;
+import com.yahoo.config.provision.Zone;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Calcalute cost for for an application instance. I.e one running vespa application in one zone.
```
Comment (ldalves) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/CostApplication.java]: remove one 'for'

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/restapi/CostJsonModelAdapter.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        object.setString("flavor", clusterCost.getClusterInfo().getClusterType().name());
+        object.setLong("waste", (int)clusterCost.getWaste());
+        object.setString("type", clusterCost.getClusterInfo().getClusterType().name());
+        Cursor utilObject = object.setObject("util");
+        utilObject.setDouble("cpu", clusterCost.getResultUtilization().getCpu());
+        utilObject.setDouble("mem", clusterCost.getResultUtilization().getMemory());
+        utilObject.setDouble("disk", clusterCost.getResultUtilization().getDisk());
+        utilObject.setDouble("diskBusy", clusterCost.getResultUtilization().getDiskBusy());
+        Cursor usageObject = object.setObject("usage");
+        usageObject.setDouble("cpu", clusterCost.getSystemUtilization().getCpu() * 100);
```
Comment (ldalves) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/restapi/CostJsonModelAdapter.java]: I guess we don't need * 100 here, do we? The metric from getSystemUtilization() is already computed with * 100.

**Review by ldalves — APPROVED** [review]

**Comment by bratseth** [comment]
I think this should be completely integrated into the application model instead of being done as something on the side - some of TODO's alluded to that but now they are removed?

I think this should be the end state:
- Cost and utilization are additional values returned for an Application in application/v4
- A maintenance job fetches the node info of each application and updates the application with it
(where it is persisted as regular data of the Application).
- Another maintenance job does the same with utilization data.
- Application has methods (in subobjects, if cleaner) that computes the cost and utilization info on the fly from the node and utilization data which is always present (due to the above).

Advantages of this are much less code, a simpler model where everything is done the same way, and that we can then use util and node info also for other purposes that computing these values.

**Comment by smorgrav** [comment]
@bratseth The TODOs was mostly complied with and some where considered :) 

The only use case we have for this today is to return utilisations for all applications in all zones. Putting this into the application api sounds like optimising for a use case we don't have. I doubt that the code base would be significantly smaller (added maintainers, dashboard changes...  and the computations would be the same). I also like the decoupling when done on the side (one thing that does one thing). 

But I have no problem with putting it into the Application module - less wiring is always good. 

If you still are convinced that that is the way to do it - I'll do that tomorrow.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/restapi/CostJsonModelAdapter.java (hunk 3 above)
Comment (smorgrav) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/cost/restapi/CostJsonModelAdapter.java]: I went with ratios here to make the calculations easier, leaving percentage for the 'presentation'

**Comment by bratseth** [comment]
> I also like the decoupling when done on the side (one thing that does one thing).

Decoupling as vertical silos, where each aspect of a thing lives in a separate world.
The problems I have with that approach is that there is no accumulation of competence/capability - each new aspects is as much work (and code) as any other has been before, and that it does not give us a set of concept useful for understanding, thinking and communicating (i.e there is no such thing as an "application" ... only various aspects of applications - "postmodernist design"? :-) )

**Comment by smorgrav** [comment]
I take that as a yes :)

**Comment by bratseth** [comment]
I'm more trying to convince you. That is more important than this particular thing.  :-)

**Comment by smorgrav** [comment]
@bratseth Final thoughts? This PR needs to go in together with the internal PR 4766 implementing the new method for the configserver client.

**Comment by bratseth** [comment]
Excellent - let's get it merged.

**Pull request merged**